### DISCUSSION TITLE
Publishing an IOS app (capacitor)

### DISCUSSION THREAD

**Discussion by jewells77** [codemagic-ci-cd/codemagic-docs#1977] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Ionic/Cordova

### Steps to reproduce

This is my codemagic.yaml, and I'm facing an issue with the step "- name: Build ipa for distribution".

```
workflows:
  ionic-capacitor-ios-workflow:
    name: Ionic Capacitor iOS Workflow
    max_build_duration: 120
    instance_type: mac_mini_m1
    integrations:
      app_store_connect: Codemagic API Key
    environment:
      ios_signing:
        # distribution_type: app_store
        bundle_identifier: org.hireaspy.app
        provisioning_profiles:
          - codemagic_provisioning_profiles
        certificates:
          - Development_Certificate
      vars:
        # Ionic Xcode worskspace and scheme
        XCODE_WORKSPACE: "ios/App/App.xcworkspace"
        XCODE_SCHEME: "App"
        APP_STORE_APP_ID: 6456386785 # <-- Put the app id number here. This is found in App Store Connect > App > General > App Information
      node: 16.15.1
      xcode: latest
      cocoapods: 1.12.1
    triggering:
      events:
        - push
      branch_patterns:
        - pattern: develop
          include: true
          source: true
    scripts:
      - name: Install npm dependencies for Ionic project
        script: |
          npm install
      - name: Cocoapods installation
        script: |
          cd ios/App && pod install
      - name: Update dependencies and copy web assets to native project
        script: |
          npx cap sync ios # <- update native dependencies and copy web assets to native project
          npx cap copy ios # <- use this is you don't need to update native dependencies
      - name: Set up code signing settings on Xcode project
        script: |
          xcode-project use-profiles
      - name: Increment build number
        script: |
          cd $CM_BUILD_DIR/ios/App
          LATEST_BUILD_NUMBER=$(app-store-connect get-latest-app-store-build-number "$APP_ID")
          agvtool new-version -all $(($LATEST_BUILD_NUMBER + 1))
      - name: Build ipa for distribution
        script: |
          xcode-project build-ipa \
            --workspace "$CM_BUILD_DIR/$XCODE_WORKSPACE" \
            --scheme "$XCODE_SCHEME"
    artifacts:
      - build/ios/ipa/*.ipa
      - /tmp/xcodebuild_logs/*.log
      - $HOME/Library/Developer/Xcode/DerivedData/**/Build/**/*.app
      - $HOME/Library/Developer/Xcode/DerivedData/**/Build/**/*.dSYM
    publishing:
      email:
        recipients:
          - john.doe@gmail.com
        notify:
          success: false     # To not receive a notification when a build succeeds
          failure: false     # To not receive a notification when a build fails
      app_store_connect:
        auth: integration

        # Configuration related to TestFlight (optional)
        # Note: This action is performed during post-processing.
        submit_to_testflight: true
        # beta_groups: # Specify the names of beta tester groups that will get access to the build once it has passed beta review.
          # - group name 1
          # - group name 2

        # Configuration related to App Store (optional)
        # Note: This action is performed during post-processing.
        submit_to_app_store: false
```****

### Expected results

It should publish an iOS app.

### Actual results

 It is not able to progress beyond this step.

![rd8wt](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/75bc7b9a-14b0-4157-b8df-93b196f9c0e5)


### Build id (optional)

64df04fe25abe981552a105a

**Comment by himesh-cm** [comment]
Hi @jewells77,

Error says 
```
error: No signing certificate "iOS Distribution" found: No "iOS Distribution" signing certificate matching team ID "9P2HJFFNT5" with a private key was found. (in target 'App' from project 'App')
```
The profile you uploaded does not have the Distribution certificate.
Please take a look at the image I've included below. You need to choose the certificate which you uploaded In your `Teams > Settings> Code signing identities > Profiles`.
 
<img width="1346" alt="Screenshot 2023-01-31 at 2 39 07 PM" src="https://github.com/codemagic-ci-cd/codemagic-docs/assets/106069365/9145cc08-0f2c-4588-a1c4-c1c1d86cfa5b">

**Reply by jewells77** [reply]
It shows the same error. When I tried to "Get certificates from Apple Developer Portal", it showed "Unavailable certificates", so I created it from "Generate a new code signing certificate", and I added (IOS Distribution) to the profile, and it shows "Certificate Not uploaded X"

![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/1f06d21a-2451-406a-ad93-6e916f894174)

**Reply by himesh-cm** [reply]
Your profile includes the `new certificate` per the screenshot I shared in my last reply?

**Reply by jewells77** [reply]
@himesh-cm yes, 
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/9f773758-bf36-4079-9c0b-655a9f4f90e1)

**Reply by himesh-cm** [reply]
Please download the profile from Codemagic, upload it again, and run the build. Please inform me of the outcome.

**Reply by jewells77** [reply]
I have downloaded the profile and uploaded it again, and it caused the same error on the build step "Build ipa for distribution".

```
Using Xcode 14.3.1 (14E300c)
Check App.xcworkspace build settings
Execute "xcodebuild -workspace /Users/builder/clone/ios/App/App.xcworkspace -scheme App -showBuildSettings"

Archive App.xcworkspace
Execute "xcodebuild -workspace /Users/builder/clone/ios/App/App.xcworkspace -scheme App -archivePath build/ios/xcarchive/App_xfwi88lv.xcarchive archive COMPILER_INDEX_STORE_ENABLE=NO DEVELOPMENT_TEAM=1234567890 'CODE_SIGN_IDENTITY=iPhone Distribution' CODE_SIGN_STYLE=Manual"

    Run script build phase '[CP] Embed Pods Frameworks' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase. (in target 'App' from project 'App')
Failed to archive /Users/builder/clone/ios/App/App.xcworkspace


Build failed :|
Step 8 script `Build ipa for distribution` exited with status code 65
```

**Reply by himesh-cm** [reply]
@jewells77 , Please check the script for the build ipa command as mentioned [here](https://github.com/codemagic-ci-cd/codemagic-sample-projects/blob/main/ionic/ionic-capacitor-demo-project/codemagic.yaml#L48) in the sample project and re-run it.

**Reply by jewells77** [reply]
I have changed the build script:
```
      - name: Build ipa for distribution
        script: |
          cd $CM_BUILD_DIR/ios/App
          xcode-project build-ipa \
            --workspace "$XCODE_WORKSPACE" \
            --scheme "$XCODE_SCHEME"
```

But It shows an error: the path does not exist.

```
usage: xcode-project build-ipa [-h] [--log-stream {stderr,stdout}] [--no-color] [--version] [-s] [-v] [--project XCODE_PROJECT_PATH] [--workspace XCODE_WORKSPACE_PATH] [--target TARGET_NAME] [--config CONFIGURATION_NAME] [--scheme SCHEME_NAME] [--clean] [--no-show-build-settings] [--archive-directory ARCHIVE_DIRECTORY] [--archive-flags ARCHIVE_FLAGS] [--archive-xcargs ARCHIVE_XCARGS] [--ipa-directory IPA_DIRECTORY] [--export-options-plist EXPORT_OPTIONS_PLIST] [--export-flags EXPORT_FLAGS] [--export-xcargs EXPORT_XCARGS] [--remove-xcarchive] [--disable-xcpretty] [--xcpretty-options XCPRETTY_OPTIONS]
xcode-project build-ipa: error: argument --workspace: Path "ios/App/App.xcworkspace" does not exist


Build failed :|
Step 8 script `Build ipa for distribution` exited with status code 2
```


![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/23b88829-e36e-4beb-a8fe-f98f92224b22)

**Reply by himesh-cm** [reply]
Hi @jewells77, Please [ssh](https://docs.codemagic.io/troubleshooting/accessing-builder-machine-via-ssh/) into the build machine and check if the path is correct, and try and run the command via ssh terminal for debugging purposes.

**Reply by jewells77** [reply]
I can see the path "ios/App/App.xcworkspace"

![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/47d16ca2-15b1-4329-855c-c21ce03e621c)

I ran the command and got an error, and it is still the same.
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/6139a039-21e3-4311-a3e4-b640dca1f6b0)

**Reply by himesh-cm** [reply]
Can you please share the build id for the error you mentioned [here](https://github.com/orgs/codemagic-ci-cd/discussions/1977#discussioncomment-6799845)?

**Reply by jewells77** [reply]
Build Id: 64e7452b797fc5e6762859de

**Reply by himesh-cm** [reply]
remove this command `cd $CM_BUILD_DIR/ios/App` and try re-running the build

**Reply by jewells77** [reply]
Error:
```
Using Xcode 14.3.1 (14E300c)
Check App.xcworkspace build settings
Execute "xcodebuild -workspace ios/App/App.xcworkspace -scheme App -showBuildSettings"

Archive App.xcworkspace
Execute "xcodebuild -workspace ios/App/App.xcworkspace -scheme App -archivePath build/ios/xcarchive/App_28svdfpu.xcarchive archive COMPILER_INDEX_STORE_ENABLE=NO DEVELOPMENT_TEAM=1234567890 'CODE_SIGN_IDENTITY=iPhone Distribution' CODE_SIGN_STYLE=Manual"

    Run script build phase '[CP] Embed Pods Frameworks' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase. (in target 'App' from project 'App')
Failed to archive ios/App/App.xcworkspace


Build failed :|
Step 8 script `Build ipa for distribution` exited with status code 65
```

**Reply by himesh-cm** [reply]
Build id, please.

**Reply by jewells77** [reply]
build id: 64ed7b813a8fd674c874b9d2

**Reply by himesh-cm** [reply]
The errors from the logs are 

`error: No signing certificate "iOS Distribution" found: No "iOS Distribution" signing certificate matching team ID "9P2HJFFNT5" with a private key was found. (in target 'App' from project 'App')`

Please check if you have uploaded a distribution certificate under `Teams > Settings > Code signing identities`
You should also see a `x` against the uploaded profile.

**Reply by jewells77** [reply]
I have already done that, as you said above. https://github.com/orgs/codemagic-ci-cd/discussions/1977#discussioncomment-6760409 , https://github.com/orgs/codemagic-ci-cd/discussions/1977#discussioncomment-6787580

**Reply by himesh-cm** [reply]
Based on the screenshot you shared, it appears that there is an issue with the certificate you uploaded as indicated by the red cross. To successfully publish the app, you will need a `distribution` certificate. Have you checked if you have one under the `Certificates` tab?

**Reply by jewells77** [reply]
ok with ```distribution``` certificate. I got a green tick
![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/53d79934-0dbf-44b3-ad4c-28951f0ccb09)

But it failed in the build step "Build ipa for distribution"

```
Failed to archive ios/App/App.xcworkspace. The following build commands failed:

CpResource /Users/builder/Library/Developer/Xcode/DerivedData/App-dfqlwfgrvlmjunbkfzimjqcghcbo/Build/Intermediates.noindex/ArchiveIntermediates/App/InstallationBuildProductsLocation/Applications/App.app/public /Users/builder/clone/ios/App/App/public (in target 'App' from project 'App')
	...
	CompileStoryboard /Users/builder/clone/ios/App/App/Base.lproj/Main.storyboard (in target 'App' from project 'App')
	cd /Users/builder/clone/ios/App
	/Applications/Xcode-14.3.app/Contents/Developer/usr/bin/ibtool --errors --warnings --notices --module App --output-partial-info-plist /Users/builder/Library/Developer/Xcode/DerivedData/App-dfqlwfgrvlmjunbkfzimjqcghcbo/Build/Intermediates.noindex/ArchiveIntermediates/App/IntermediateBuildFilesPath/App.build/Release-iphoneos/App.build/Base.lproj/Main-SBPartialInfo.plist --auto-activate-custom-fonts --target-device iphone --target-device ipad --minimum-deployment-target 12.0 --output-format human-readable-text --compilation-directory /Users/builder/Library/Developer/Xcode/DerivedData/App-dfqlwfgrvlmjunbkfzimjqcghcbo/Build/Intermediates.noindex/ArchiveIntermediates/App/IntermediateBuildFilesPath/App.build/Release-iphoneos/App.build/Base.lproj /Users/builder/clone/ios/App/App/Base.lproj/Main.storyboard
	CompileStoryboard /Users/builder/clone/ios/App/App/Base.lproj/LaunchScreen.storyboard (in target 'App' from project 'App')
	cd /Users/builder/clone/ios/App
	/Applications/Xcode-14.3.app/Contents/Developer/usr/bin/ibtool --errors --warnings --notices --module App --output-partial-info-plist /Users/builder/Library/Developer/Xcode/DerivedData/App-dfqlwfgrvlmjunbkfzimjqcghcbo/Build/Intermediates.noindex/ArchiveIntermediates/App/IntermediateBuildFilesPath/App.build/Release-iphoneos/App.build/Base.lproj/LaunchScreen-SBPartialInfo.plist --auto-activate-custom-fonts --target-device iphone --target-device ipad --minimum-deployment-target 12.0 --output-format human-readable-text --compilation-directory /Users/builder/Library/Developer/Xcode/DerivedData/App-dfqlwfgrvlmjunbkfzimjqcghcbo/Build/Intermediates.noindex/ArchiveIntermediates/App/IntermediateBuildFilesPath/App.build/Release-iphoneos/App.build/Base.lproj /Users/builder/clone/ios/App/App/Base.lproj/LaunchScreen.storyboard
```

build id: 64f2f37721f8a386db62f1f1

**Reply by himesh-cm** [reply]
Yes, code signing related errors seems to be resolved for now.
The issue is with the path
`error: /Users/builder/clone/ios/App/App/public: No such file or directory (in target 'App' from project 'App')`
I suggest, you `ssh` into the machine and run the commands to see what is the correct path, depending upon your project structure.

**Reply by himesh-cm** [reply]
Hi @jewells77, Did you get a chance to debug the issue via `ssh`?

**Reply by jewells77** [reply]
Okay, I have fixed it, but it took a long time in the last step "Cleaning up" and then failed to "Publishing failed :| Failed to publish App.ipa to App Store Connect." Build id: 6502926c52f3bee3cbcce0a3

![image](https://github.com/codemagic-ci-cd/codemagic-docs/assets/79835068/2991b377-6b32-4c62-a3a4-48e47c008ebf)

**Reply by himesh-cm** [reply]
Hi @jewells77, as the error message under `publishing` section suggests, `A required agreement is missing or has expired.` please accept app store agreements [here](https://appstoreconnect.apple.com/agreements/#/) and re-run the build

**Reply by himesh-cm** [reply]
Hi @jewells77, While I’m closing our discussion for now, if you have any other questions or suggestions, feel free to reply back or start a new discussion anytime. Have a nice day!

**Reply by jewells77** [reply]
I was fixing some information in the app store and code. I will get back to you if there are any blockers. Thank you very much.
